Packs/ApiModules/Scripts/TAXII2ApiModule/TAXII2ApiModule.py

```diff
@@ -0,0 +1,420 @@
+
+STIX_2_TYPES_TO_CORTEX_TYPES = {
+    "ipv4-addr": FeedIndicatorType.IP,
+    "ipv6-addr": FeedIndicatorType.IPv6,
+    "domain": FeedIndicatorType.Domain,
+    "domain-name": FeedIndicatorType.Domain,
+    "url": FeedIndicatorType.URL,
+    "md5": FeedIndicatorType.File,
+    "sha-1": FeedIndicatorType.File,
+    "sha-256": FeedIndicatorType.File,
+    "file:hashes": FeedIndicatorType.File,
+}
+
+STIX_2_TYPES_TO_CORTEX_CIDR_TYPES = {
+    "ipv4-addr": FeedIndicatorType.CIDR,
+    "ipv6-addr": FeedIndicatorType.IPv6CIDR,
+}
+
+
+class Taxii2FeedClient:
+    def __init__(
+        self,
+        url,
+        collection_to_fetch,
+        proxies,
+        verify,
+        username=None,
+        password=None,
+        field_map=None,
+    ):
+        """
+        TAXII 2 Client used to poll and parse indicators in XSOAR formar
+        :param url: discovery service URL
+        :param collection_to_fetch: Collection to fetch objects from
+        :param proxies: proxies used in request
+        :param verify: verify https
+        :param username: username used for basic authentication OR api_key used for authentication
+        :param password: password used for basic authentication
+        :param field_map:
+        """
+        self.server = None
+        self.api_root = None
+        self.collections = None
+
+        self.collection_to_fetch = collection_to_fetch
+        self.base_url = url
+        self.proxies = proxies
+        self.verify = verify
+        self.auth = None
+        if username and password:
+            if username == DFLT_API_USERNAME:
+                self.auth = TokenAuth(key=password)
+            else:
+                self.auth = requests.auth.HTTPBasicAuth(username, password)
+
+        self.field_map = field_map if field_map else {}
+        self.indicator_regexes = [
+            re.compile(INDICATOR_EQUALS_VAL_PATTERN),
+            re.compile(HASHES_EQUALS_VAL_PATTERN),
+        ]
+        self.cidr_regexes = [
+            re.compile(CIDR_ISSUBSET_VAL_PATTERN),
+            re.compile(CIDR_ISUPPERSET_VAL_PATTERN),
+        ]
+
+    def init_server(self, version=TAXII_VER_2_0):
+        """
+        Initializes a server in the requested version
+        :param version: taxii version key (either 2.0 or 2.1)
+        """
+        server_url = urljoin(self.base_url)
+        if version is TAXII_VER_2_0:
+            self.server = v20.Server(
+                server_url, verify=self.verify, auth=self.auth, proxies=self.proxies,
+            )
+        else:
+            self.server = v21.Server(
+                server_url, verify=self.verify, auth=self.auth, proxies=self.proxies,
+            )
+
+    def init_roots(self):
+        """
+        Initializes the api roots (used to get taxii server objects)
+        """
+        if not self.server:
+            raise ValueError(
+                "Please init server prior to calling init_roots. Cannot init roots without server."
```

Comment:
what does the customer can do upon seeing this error?
let's try to raise an action item..

Reply:
I can remove this error completely. It was meant for devs, but I can see the code can gracefully recover by calling `self.init_server()`